File: skills/webhook-consumer-architecture/examples/go/net-http/observability.go
package examples

import (
	"context"
	"time"
)

func (w *Worker) ProcessWithObservability(ctx context.Context, deliveryID string) error {
	startedAt := time.Now()

	delivery, err := w.DeliveryStore.Get(ctx, deliveryID)
	if err != nil {
		return err
	}

	w.Logger.Info("webhook delivery received",
		"provider", delivery.Provider,
		"delivery_id", delivery.ID,
		"event_id", delivery.EventID,
	)

	err = w.Pipeline.ProcessDelivery(ctx, delivery)
	w.Metrics.ObserveDuration("webhook.processing.ms", time.Since(startedAt))

	if err == nil {
		w.Metrics.Increment("webhook.delivery.completed", delivery.Provider)
	}

	return err
}

File: skills/webhook-consumer-architecture/examples/python/django/observability.py
import logging
import time

logger = logging.getLogger("webhooks")


def process_with_observability(delivery):
    started_at = time.monotonic()

    logger.info(
        "webhook.delivery.received",
        extra={
            "provider": delivery.provider,
            "delivery_id": delivery.id,
            "event_id": delivery.event_id,
        },
    )

    webhook_pipeline.process_delivery(delivery)

    metrics.timing("webhook.processing.ms", (time.monotonic() - started_at) * 1000)
    metrics.increment("webhook.delivery.completed", tags={"provider": delivery.provider})
